src/Raven.Server/Documents/Handlers/StreamCsvDocumentQueryResultWriter.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using CsvHelper;
+using Microsoft.AspNetCore.Http;
+using Raven.Server.Documents.Queries;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.Handlers
+{
+    internal class StreamCsvDocumentQueryResultWriter : IStreamDocumentQueryResultWriter
+    {
+        private HttpResponse _response;
+        private DocumentsOperationContext _context;
+        private StreamWriter _writer;
+        private CsvWriter _csvWriter;
+        private string[] _properties;
+
+        public StreamCsvDocumentQueryResultWriter(HttpResponse response, Stream stream, DocumentsOperationContext context, string[] properties = null)
+        {
+            _response = response;
+            _writer = new StreamWriter(stream, Encoding.UTF8);
+            _csvWriter = new CsvWriter(_writer);
+            _context = context;
+            _properties = properties;
+        }
+
+        public void Dispose()
+        {
+            _csvWriter?.Dispose();
+            _writer?.Dispose();
+        }
+
+        public void StartResponse()
+        {
+        }
+
+        public void StartResults()
+        {
+        }
+
+        public void EndResults()
+        {
+        }
+
+        public void AddResult(Document res)
+        {
+            if (_properties == null)
+            {
+                GetPropertiesAndWriteCsvHeader(res);
+            }
+            foreach (var property in _properties)
+            {
+                object r;
+                //TODO: verify that all blittable property types implement ToString
```

Comment:
They don't, they can be complex objects, arrays, etc.
The properties you use need to be blittable paths.

Reply:
yes, i assumed more work needs to be done around here.

Reply:
You output the error as is, which will corrupt the CSV file, but it will give us an error that they can see

Reply:
what should be the path of an object that reside inside an array? Foo.Bar[5].Name? and what if no other document have an object under Bar[5]? in 3.x arrays seems to be simply written as strings